compiler/lib/Dialect/mhlo/Transforms/HloSimplify.cpp

```diff
@@ -132,23 +151,107 @@ struct SimplifyDotGeneralToBroadcastMultiply
       return failure();
     }
 
-    // k == 1
-    if (lhsType.getDimSize(dimensionNumbers.getLhsContractingDimensions()[0]) !=
-            1 ||
-        rhsType.getDimSize(dimensionNumbers.getRhsContractingDimensions()[0]) !=
-            1) {
-      return failure();
+    auto is_single_k = [&]() {
+      return lhsType.getDimSize(
+                 dimensionNumbers.getLhsContractingDimensions()[0]) == 1 &&
+             rhsType.getDimSize(
+                 dimensionNumbers.getRhsContractingDimensions()[0]) == 1;
+    };
+
+    auto is_single_m_n = [&]() {
+      return lhsType.getDimSize(lhsType.getRank() - 2) == 1 ||
+             rhsType.getDimSize(rhsType.getRank() - 1) == 1;
+    };
+
+    auto resultRank = op.getType().getRank();
+    auto resultShape = op.getType().getShape();
+    auto resultElementTy = op.getType().getElementType();
+    auto has_batch = dimensionNumbers.getLhsBatchingDimensions().size() != 0 ||
+                     dimensionNumbers.getRhsBatchingDimensions().size();
+    auto B = has_batch ? resultShape[0] : 0;
+    auto M = lhsType.getDimSize(lhsType.getRank() - 2);
+    auto N = rhsType.getDimSize(rhsType.getRank() - 1);
+    auto LK =
+        lhsType.getDimSize(dimensionNumbers.getLhsContractingDimensions()[0]);
+    auto RK =
+        rhsType.getDimSize(dimensionNumbers.getRhsContractingDimensions()[0]);
+    auto K = std::max(LK, RK);
+
+    if (is_single_k()) {
+      // k == 1
+      auto iota = llvm::iota_range<int64_t>(0, lhsType.getRank(), false);
+      llvm::SmallVector<int64_t> bcastDims(iota.begin(), iota.end());
+      Value bcastLhs = rewriter.create<mhlo::BroadcastInDimOp>(
+          loc, op.getType(), op.getLhs(), rewriter.getI64TensorAttr(bcastDims));
+      Value bcastRhs = rewriter.create<mhlo::BroadcastInDimOp>(
+          loc, op.getType(), op.getRhs(), rewriter.getI64TensorAttr(bcastDims));
+      rewriter.replaceOpWithNewOp<mhlo::MulOp>(op, op.getType(), bcastLhs,
+                                               bcastRhs);
+      return success();
+    } else if (is_single_m_n()) {
+      // m == 1 or n == 1
+      const bool is_single_m = M == 1;
+      auto lhsRank = lhsType.getRank();
+      auto rhsRank = rhsType.getRank();
+
+      // broadcast lhs and rhs.
+      auto lIota = llvm::iota_range<int64_t>(0, lhsRank, false);
+      auto rIota = llvm::iota_range<int64_t>(0, rhsRank, false);
+      llvm::SmallVector<int64_t, 3> lhsBcastDims(lIota.begin(), lIota.end());
```

Comment:
I found that `lhsBcastDims = llvm::to_vector(llvm::seq<int64_t>(0, lshRank))` is more simple.

Reply:
> I found that `lhsBcastDims = llvm::to_vector(llvm::seq<int64_t>(0, lshRank))` is more simple.

looks good!